### DISCUSSION TITLE
Variable assignments in System.Random's xorshiro256ss' NextUInt64() can be removed

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#95116] [General] [answered]
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs,2cd982a09ee3ad05,references

![image](https://github.com/dotnet/runtime/assets/35312807/0b1a7d2c-4631-47f1-9ec4-09421f747309)

In particular, these assignments doesn't have to be there, the cpu is copying bytes which I say is ~30% overhead after seeing in profiler that's how much it was spending on dry run of the function

**Comment by huoyaoyuan** [comment, marked as answer]
Simple benchmark does not support this:

```csharp
using System.Numerics;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Program
{
    static void Main()
    {
        BenchmarkRunner.Run<Program>();
    }

    private ulong _s0 = 123, _s1 = 456, _s2 = 789, _s3 = 1000;

    [Benchmark(Baseline = true)]
    public ulong NextUInt64()
    {
        ulong s0 = _s0, s1 = _s1, s2 = _s2, s3 = _s3;

        ulong result = BitOperations.RotateLeft(s1 * 5, 7) * 9;
        ulong t = s1 << 17;

        s2 ^= s0;
        s3 ^= s1;
        s1 ^= s2;
        s0 ^= s3;

        s2 ^= t;
        s3 = BitOperations.RotateLeft(s3, 45);

        _s0 = s0;
        _s1 = s1;
        _s2 = s2;
        _s3 = s3;

        return result;
    }

    [Benchmark]
    public ulong NextUInt64Alt()
    {
        ulong result = BitOperations.RotateLeft(_s1 * 5, 7) * 9;
        ulong t = _s1 << 17;

        _s2 ^= _s0;
        _s3 ^= _s1;
        _s1 ^= _s2;
        _s0 ^= _s3;

        _s2 ^= t;
        _s3 = BitOperations.RotateLeft(_s3, 45);

        return result;
    }
}
```

BenchmarkDotNet v0.13.7, Windows 11 (10.0.22631.2715)
Intel Core i9-10900X CPU 3.70GHz, 1 CPU, 20 logical and 10 physical cores
.NET SDK 8.0.100
  [Host]     : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2


|        Method |      Mean |     Error |    StdDev | Ratio | RatioSD |
|-------------- |----------:|----------:|----------:|------:|--------:|
|    NextUInt64 | 0.6454 ns | 0.0094 ns | 0.0083 ns |  1.00 |    0.00 |
| NextUInt64Alt | 0.8203 ns | 0.0276 ns | 0.0245 ns |  1.27 |    0.04 |

**Comment by hez2010** [comment]
Actually it's the opposite, by assigning those fields to locals the JIT can assume no other code can alter the values, which can result in better codegen.

**Comment by Xyncgas** [comment]
I have a benchmark as well
![image](https://github.com/dotnet/runtime/assets/35312807/ce4f9775-48b3-488b-af56-7dcb5702168c)
Microsoft Visual Studio Community 2022 (64-bit) - Preview
Version 17.9.0 Preview 1.0
laptop plugged in

Although it's interesting I believe I am wrong because the Visual Studio profiler doesn't look right to me

**Reply by hez2010** [reply]
The Visual Studio profiler is correct because the assignment here involves one-time memory access which seems to be the most expensive operation of the whole function.
But if you remove those assignments, a memory access will happen on every state access which will result in lower performance.

**Reply by Xyncgas** [reply]
Thank u for your wisdom

**Reply by Xyncgas** [reply]
why is it expensive to read {s1,s2,s3,s4}, shouldn't the entire rng be in cache it's only 32 bytes, in the profiling it's only a while loop calling NextUInt64()

**Reply by huoyaoyuan** [reply]
Addressing in cache is still expensive comparing to registers. Further more, some intermidiate results can be eliminated by using additional register.
